### PR TITLE
[bf #331] raising exceptions when poly ill defined

### PR DESCRIPTION
**Pull request by lasofivec**
As discussed in #331 : 
The polygon needs to be properly defined in order for the algorithm `isClockwise` works correctly.
If it finds an issue (null volume) it will raise an exception.

The exception is caught and raised whenever function used
**Warning**: I didn't manage to properly catch and raise the error.  Can you see what I am doing wrong ? I left a `print` for the moment

I also realized in _core, `isClockwise` is defined sometimes **before** checking if the polygon is closed, however the algorithm needs for the polygon to be closed.

Lastly: some flake8 changes

### CODE REVIEW

File: tofu/geom/_GG.pyx [hunk 1]
```diff
@@ -206,12 +204,23 @@ def Poly_isClockwise(np.ndarray[double,ndim=2] Poly):
...
           mvx[idp1]  * (mvy[idm1]  - mvy[idmin])
+    if abs(res) < _VSMALL:
+        err_msg += ("In Poly_isClockwise : \n"
+                    + "   Found lowest right point at index : "
+                    + str(idmin)
+                    + ", of coordinates :" + str(mvx[idmin])
+                    + ", " + str(mvy[idmin]) + ".\n"
+                    + "   The two neighboring points are : "
+                    + str(idm1) + " and " + str(idp1) + ".")
+        print(err_msg)
```
Comment (lasofivec) [inline tofu/geom/_GG.pyx]: added this since I didnt manage to catch the Exception later

File: tofu/geom/_GG.pyx [hunk 2]
```diff
@@ -274,8 +283,11 @@ def Poly_Order(np.ndarray[double,ndim=2] Poly, str order='C', Clock=False,
     if not np.allclose(poly[:,0],poly[:,-1], atol=_VSMALL):
         poly = np.concatenate((poly,poly[:,0:1]),axis=1)
     if poly.shape[0]==2 and not Clock is None:
-        if not Clock==Poly_isClockwise(poly):
-            poly = poly[:,::-1]
+        try:
+            if not Clock==Poly_isClockwise(poly):
+                poly = poly[:,::-1]
+        except Exception as excp:
+            raise excp
```
Comment (lasofivec) [inline tofu/geom/_GG.pyx]: This is where is called when creating the config. Tried several things...

File: tofu/geom/_comp.py [hunk 3]
```diff
@@ -34,9 +34,12 @@ def _Struct_set_Poly(
...
     # Make Poly closed, counter-clockwise, with '(cc,N)' layout and arrayorder
-    Poly = _GG.Poly_Order(
-        Poly, order="C", Clock=False, close=True, layout="(cc,N)", Test=True
-    )
+    try:
+        Poly = _GG.Poly_Order(
+            Poly, order="C", Clock=False, close=True, layout="(cc,N)", Test=True
+            )
+    except Exception as excp:
+        print(excp)
```
Comment (lasofivec) [inline tofu/geom/_comp.py]: and trying to catch and print the exception message...

File: tofu/geom/_core.py [hunk 4]
```diff
@@ -5574,17 +5574,17 @@ def _kInOut_Isoflux_inputs_usr(self, lPoly, lVIn=None):
         if type(lPoly) is list:
             for ii in range(nPoly):
                 # Check closed and anti-clockwise
-                if _GG.Poly_isClockwise(lPoly[ii]):
-                    lPoly[ii] = lPoly[ii][:, ::-1]
                 if not np.allclose(lPoly[ii][:, 0], lPoly[ii][:, -1]):
                     lPoly[ii] = np.concatenate(
                         (lPoly[ii], lPoly[ii][:, 0:1]), axis=-1
                     )
```
Comment (lasofivec) [inline tofu/geom/_core.py]: first we close the poly then we check if clockwise

File: tofu/geom/_core.py [hunk 5]
```diff
@@ -6900,7 +6906,7 @@ def get_summary(
         # Prepare
         kout = self._dgeom["kOut"]
         indout = self._dgeom["indout"]
-        lS = self._dconfig["Config"].lStruct
+        # lS = self._dconfig["Config"].lStruct
```
Comment (lasofivec) [inline tofu/geom/_core.py]: is not being used, so I commented it

File: tofu/geom/_basic_geom_tools.pxd [hunk 6]
```diff
@@ -1,6 +1,7 @@
 # cython: boundscheck=False
 # cython: wraparound=False
 # cython: cdivision=True
+# cython: initializedcheck=False
```
Comment (Didou09) [inline tofu/geom/_basic_geom_tools.pxd]: What does this flag do ?

**Comment by lasofivec** [comment]
Ok, my bad. Forgot that it needed to be a "-2" instead of "-1". 
This solved the issue, event hough I am not entirely sure why did this also solve the issue with the plotting of the structure ?

Capture d’écran 2020-01-22 à 13 57 00

File: tofu/geom/_basic_geom_tools.pxd (hunk 6 above)
Comment (lasofivec) [reply tofu/geom/_basic_geom_tools.pxd]: Cython normally checks if a memory view is initialized before using it. This turns the verification off (for speed purposes). I wanted to match the flags with the GG file.

File: tofu/geom/_GG.pyx [hunk 7]
```diff
@@ -206,12 +204,23 @@ def Poly_isClockwise(np.ndarray[double,ndim=2] Poly):
...
+    if abs(res) < _VSMALL:
+        err_msg += ("In Poly_isClockwise : \n"
+                    + "   Found lowest right point at index : "
+                    + str(idmin)
+                    + ", of coordinates :" + str(mvx[idmin])
+                    + ", " + str(mvy[idmin]) + ".\n"
+                    + "   The two neighboring points are : "
+                    + str(idm1) + " and " + str(idp1) + ".")
+        print(err_msg)
+        raise Exception(err_msg) # not working
```
[the remark this answers is not in the document]
Comment (lasofivec) [reply tofu/geom/_GG.pyx]: it is being raised if the function is called directly called from python. For example, the test that is failing is showing:
``

**Comment by Didou09** [comment]
It's because the plotting procedure also depends on the order of the polygon (clockwise or not).
The position of the labels (indices) is determined by the in/out vectors perpendicular to each segment, and these in/out vectors may be wrong if there was a bug in one of the preceding steps.

**Comment by Didou09** [comment]
I'm testing it locally, but it seems we are fine,
I just want to double-check because of the C-contiguous error I found last time

**Comment by Didou09** [comment]
So, locally it seems we are good,
But travis finds an error in the unit tests:
```
======================================================================
ERROR: tofu.tests.tests01_geom.tests01_GG.test02_Poly_CLockOrder
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/ToFuProject/tofu/tofu/tests/tests01_geom/tests01_GG.py", line 119, in test02_Poly_CLockOrder
    not GG.Poly_isClockwise(P), P.flags['C_CONTIGUOUS'],
  File "tofu/geom/_GG.pyx", line 223, in tofu.geom._GG.Poly_isClockwise
Exception: In Poly_isClockwise : 
   Found lowest right point at index : 1, of coordinates :0.0, 0.0.
   The two neighboring points are : 0 and 0.
```

Could you have a look ?

**Comment by lasofivec** [comment]
I am hunting this bug :)

**Comment by lasofivec** [comment]
Ok it took me a while to understand the bug but it is not really a bug.....
This is what happens:

- A polygon is created in the right order: (number_of_dims, number_of_points) = (2, npts)
- It is passed to the `Poly_Order` function with the parameter `order=(N,cc)` meaning that the polygon is going to be translated so that its shape is (npts, 2)
- Then we check if the polygon is not clockwise, however in this algorithm we suppose that the polygon is in the order (2, npts).
- Original version of the algorithm (coded by you, see below) and the version in devel, the computation was done and because the algorithm will return False since it's being computed on a segment instead of the polygon. 
- Now we raise an exception if it finds that the volume is 0 (which is the case here), thus the assert `assert not GG.Poly_isClockwise(P)` fails...

So, I would suggest either: 
- Taking out the assert in the test function and maybe adding a test to check if the number of points in `GG.Poly_isClockwise` is > 2.
- Checking if we find npts < 3, and in that case translate the polygon, and let the algorithm run as is. If for example we send a polygon with shape (2,2), it will still fail (I can add in the exception message the shape of the polygon received).

I wouldn't suggest calling `GG.Poly_Order` in `GG.Poly_isClockwise` since `GG.Poly_Order` already calls `GG.Poly_isClockwise`.

FTR, original version of the function

```
@cython.cdivision(True)
@cython.wraparound(False)
@cython.boundscheck(False)
def Poly_isClockwise(cnp.ndarray[double,ndim=2] Poly):
    """ Assuming 2D closed Poly ! """
    cdef int ii, NP=Poly.shape[1]
    cdef double Sum=0.
    for ii in range(0,NP-1):
        Sum += Poly[0,ii]*Poly[1,ii+1]-Poly[0,ii+1]*Poly[1,ii]
    return Sum < 0.
 

```

**Comment by lasofivec** [comment]
I think the second solution is the most elegant, so I made changes in that direction. Let me know what you think

**Comment by Didou09** [comment]
Hum, I think I was originally translating from (2, npts) to (npts, 2) because before I was using external library Polygon2 (or Polygon3) to computed some properies (area...).
Since your optimizations, we got rid of this dependency, so we can probably stop translating from (2, npts) to (npts, 2) and stick with (2, npts) all the time.

hat should solve the issue, wouldn't it ?

**Comment by Didou09** [comment]
Also, I agree your solution is the most elegant

**Comment by lasofivec** [comment]
Yes, but this would mean we should make some major changes to PolyOrder:
- Take out the option of changing the order
- Renaming the function :)
- Checking all tests and code that call it

**Comment by lasofivec** [comment]
Probably we can open a new issue about this if you prefer

**Comment by Didou09** [comment]
Hum, good point, 

What do you think ?
I'll let you decide which option you prefer, either a hotfix (your option 2) or a full-blown rewriting of the incriminated parts.

Given that we are a bit late on time I would probably go for a hotfix, but I understand if you prefer something cleaner and more complete. your move ;-)

**Comment by lasofivec** [comment]
we can go for the hotfix (in this branch and later in devel) and open an issue for the cleaner solution.

**Pull request merged**